rpc.go

```diff
@@ -188,139 +178,120 @@ func (l *Libvirt) listen() {
+
 	switch {
-	case h.Program == constants.ProgramQEMU && h.Procedure == constants.QEMUDomainMonitorEvent:
-		streamEvent = &DomainEvent{}
+	case h.Program == constants.QEMUProgram && h.Procedure == constants.QEMUProcDomainMonitorEvent:
+		event = &DomainEvent{}
 	case h.Program == constants.Program && h.Procedure == constants.ProcDomainEventCallbackLifecycle:
-		streamEvent = &DomainEventCallbackLifecycleMsg{}
+		event = &DomainEventCallbackLifecycleMsg{}
 	}
 
-	if streamEvent != nil {
-		err := eventDecoder(buf, streamEvent)
+	if event != nil {
+		err := eventDecoder(buf, event)
 		if err != nil { // event was malformed, drop.
 			return
 		}
-		l.stream(streamEvent)
+
+		l.stream(event)
 		return
 	}
 
-	// send responses to caller
-	res := response{
-		Payload: buf,
-		Status:  h.Status,
-	}
-	l.callback(h.Serial, res)
+	// send response to caller
+	l.callback(h.Serial, response{Payload: buf, Status: h.Status})
 }
 
 // serial provides atomic access to the next sequential request serial number.
-func (l *Libvirt) serial() uint32 {
-	return atomic.AddUint32(&l.s, 1)
+func (l *Libvirt) serial() int32 {
+	return atomic.AddInt32(&l.s, 1)
 }
 
-// stream decodes domain events and sends them
-// to the respective event listener.
-func (l *Libvirt) stream(e event) error {
-	// send to event listener
-	l.em.Lock()
-	c, ok := l.events[e.GetCallbackID()]
-	l.em.Unlock()
+// stream decodes and relays domain events to their respective listener.
+func (l *Libvirt) stream(e event.Event) {
+	l.emux.RLock()
+	defer l.emux.RUnlock()
 
-	if ok {
-		// we close the channel in deregister() so that we don't block here
-		// forever without a receiver. If that happens, this write will panic.
-		defer func() {
-			recover()
-		}()
-		c.Events <- e
+	q, ok := l.events[e.GetCallbackID()]
+	if !ok {
+		return
 	}
-	return nil
-}
 
-// addStream configures the routing for an event stream.
-func (l *Libvirt) addStream(id uint32, s eventStream) {
-	l.em.Lock()
-	l.events[id] = s
-	l.em.Unlock()
+	q.Push(e)
 }
 
-// removeStream notifies the libvirt server to stop sending events
-// for the provided callback id. Upon successful de-registration the
-// callback handler is destroyed.
-func (l *Libvirt) removeStream(id uint32) error {
-	stream := l.events[id]
-	close(stream.Events)
-
-	payload := struct {
-		CallbackID uint32
-	}{
-		CallbackID: id,
-	}
+// addStream configures the routing for an event stream.
+func (l *Libvirt) addStream(s *event.Stream) {
+	l.emux.Lock()
+	defer l.emux.Unlock()
 
-	buf, err := encode(&payload)
-	if err != nil {
-		return err
-	}
+	l.events[s.CallbackID] = s
+}
 
-	_, err = l.request(stream.DeregisterProc, stream.Program, buf)
-	if err != nil {
-		return err
+// removeStream notifies the libvirt server to stop sending events for the
+// provided callback ID. Upon successful de-registration the callback handler
+// is destroyed. Subsequent calls to removeStream are idempotent and return
+// nil.
+// TODO: Fix this comment
```

Comment:
did i leave this? don't remember doing that...

Reply:
No, that was one of mine. I'll take care of it.